kobo-uncaged/kutypes.go

```diff
@@ -143,18 +144,22 @@ func (d koboDevice) CoverSize() (fullCover, libFull, libGrid image.Point) {
 		fc = image.Pt(1072, 1448)
 		lf = image.Pt(355, 479)
 		lg = image.Pt(149, 201)
-	case auraHD, auraH2O, auraH2Oed2r1, auraH2Oed2r2:
+	case auraHD, auraH2Oed2r1, auraH2Oed2r2:
 		fc = image.Pt(1080, 1440)
 		lf = image.Pt(355, 471)
 		lg = image.Pt(149, 198)
+	case auraH2O:
+		fc = image.Pt(1080, 1429)
+		lf = image.Pt(355, 473)
+		lg = image.Pt(149, 198)
 	case auraOne, auraOneLE:
 		fc = image.Pt(1404, 1872)
 		lf = image.Pt(355, 473)
 		lg = image.Pt(149, 198)
-	case forma:
+	case forma, forma32gb:
 		fc = image.Pt(1440, 1920)
-		lf = image.Pt(355, 473)
-		lg = image.Pt(149, 198)
+		lf = image.Pt(398, 530)
+		lg = image.Pt(167, 223)
```

Comment:
Are you sure about this?

If so, you might want to let davidfor know to (potentially) update the Calibre driver, as that is where I got the values.

Reply:
The sizes are at 0x63c848 in fw 4.14.12777. I'll check them to make sure.

Reply:
Here's the CFG of you want to have a look: [imagesize.pdf](https://s3.geek1011.net/temp/imagesize.pdf?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=admin%2F20190520%2F%2Fs3%2Faws4_request&X-Amz-Date=20190520T221456Z&X-Amz-Expires=604800&X-Amz-SignedHeaders=host&X-Amz-Signature=176d4baf2084a36b6ebc24932db0ccff5f0e246eaa6252391a1a538954226a2f)

Reply:
And with the sizes in decimal: [imagesizenum.pdf](https://s3.geek1011.net/temp/imagesizenum.pdf?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=admin%2F20190520%2F%2Fs3%2Faws4_request&X-Amz-Date=20190520T221735Z&X-Amz-Expires=604800&X-Amz-SignedHeaders=host&X-Amz-Signature=df30daac962ed02462ea4cec5d76528b0eba4ce12aeda7b365298b561502cc9b)

Reply:
Yes, the Calibre values are inaccurate ;).

I have a [pending](https://github.com/NiLuJe/calibre/commit/ed5478cbb547fbfa81272f084554e16068fb35bf) patch in the works.

I was just not happy with the state of dithering in both Calibre & Nickel, but things have changed a bit recently, and I can probably just port the IM ordered dither code again ;p.

Reply:
For the record, I came up with those both by feeding Nickel an ePub with a cover at the perfect AR, and by checking what it did for store-bought KePubs with more-or-less correct AR.

Reply:
I wonder how the values I found come into play though.

Reply:
I *may* also have been computing the width manually so it *actually* honors AR (and double-checking that this actually makes sense once rendered in the final widget).

Reply:
I see how that *may* have changed the x values to be *slightly* off from the actual, but can still count as *mostly* correct :smiley:.

Reply:
I obviously could only do that for devices I actually have, so, H2O & Forma ;).

Reply:
And things have probably changed since then, since a random KePub I just checked on the Forma is 355x555 for the N3_LIBRARY_FULL (it appears to be *trying* to honor the original cover's AR @ 921x1440 against a height of 555, which more or less matches what I was doing with my manual computations, except apparently with lower height bounds at the time ^^).

Reply:
I also checked that the full-screen size was often just plain wrong, Nickel's or KePubs, given the actual viewport and what Nickel does when displaying the image (that was actually my main focus at the time: getting bit-exact results from a properly encoded& dithering PNG full-screen cover).

Reply:
If what you said is correct, I misinterpreted Qt::KeepAspectRatioByExpanding for Qt::KeepAspectRatio, as what you said means it shrinks with the size being the minimum, not the maximum.

Reply:
Random one-liner to check what's on your device, for posterity's sake:

`find /mnt/onboard/.kobo-images/ -name '*N3_LIBRARY_FULL.parsed' | grep -v file___ | sed -e 's/ /\\ /g' | xargs identify`

Reply:
So, I wonder whether we should go with your calculations or try to reproduce Kobo's behaviour exactly with the values below, using bilinear scaling, and reimplementing Qt's Qt::KeepAspectRatioByExpanding mode.

Reply:
Extremely unscientific results of that:

I have a *lot* of stuff scaled around a height of x530, and quite a bit of stuff scaled around a width of 355x

Reply:
Which, hey, match the values you pulled from Nickel :D, yay?

(A lot of this is from fairly old files, though. The single KePub that ended up @ x555 is brand new, on the other hand).

Reply:
Mine are all 530 (but I normalize every book I own, so...).

Reply:
Yep, they match. *Why* do I keep posting at the exact same time as you! It's almost the 8th time today.

Reply:
@geek1011: Given the general trend, you may have been right on the Qt scaling constraints, since 355x is far more common than that single x555 ;).

Reply:
Out of my 614 thumbnails, 541 are scaled against x530, and 48 against 355x

Reply:
On a personal note, I tend to prefer Qt::KeepAspectRatioByExpanding's behavior, and IIRC, the actual widget has a crapload of leeway, so this should never be an issue.

Reply:
I think I prefer it too. I propose to switch to the values from the firmware below, and reproduce `Qt::KeepAspectRatioByExpanding` in the resize code.

Reply:
@shermp?

Reply:
Fine by me.

I'm not as picky as @NiLuJe is when it comes to cover/thumbnails quality. So doesn't really bother me what approach is taken.

Reply:
OK. If NiLuJe wants to implement this, he can, or else I'll do it later tonight or tomorrow.

Reply:
I'll leave the actual Go heavy lifting to you guys, as I'm mostly a tourist in that area ;).

Reply:
FWIW, because maths, here's Qt's logic: https://github.com/qt/qtbase/blob/d3b887707d8c2b342153f962814bb0ce2d0db2b7/src/corelib/tools/qsize.cpp

Reply:
Go's not that hard @NiLuJe! You really should learn it.

Reply:
I'm good with C & Python for now, but it is fairly non-alien looking, so there's a vague chance I could reasonably get into it one day ;).

(I have, err, fairly harsh initial gut-feeling reactions to a language's syntax :D. I'm still queasy about Lua in that respect, despite years of messing about with WoW UI & KOReader ;)).

Reply:
When I first started learning Go, I found many aspects of it more like C than I've found in most other languages I've used. Made it pretty comfortable to start producing something in a couple of days.

Reply:
> (I have, err, fairly harsh initial gut-feeling reactions to a language's syntax :D. I'm still queasy about Lua in that respect, despite years of messing about with WoW UI & KOReader ;))
I still can't get used to Lua. I've attempted to try multiple times, but given up each time. I like C so far (I've only done minor things in it, nothing big except for some embedded stuff), but I find the error handling and string stuff a bit convoluted and tedious (and the lack of portability in libraries drives me nuts too).

I'm mainly a Go, Python, C# (not as much now that I've been using Linux for around 6 years now).